fennel/testing/execute_aggregation.py

```diff
@@ -422,58 +458,79 @@ def preprocess_df_continuous_window(
 def preprocess_df_session_window(
     input_df: pd.DataFrame,
     window: Session,
-    key_fields: List[str],
+    key_fields_without_window: List[str],
     ts_field: str,
+    is_window_key_field: bool = False,
 ) -> pd.DataFrame:
     gap = window.gap_total_seconds()
-    cols = input_df.columns
+    cols = list(input_df.columns)
+    if is_window_key_field:
+        cols.append(WINDOW_FIELD_NAME)
+
+    # Get session end timestamps and if is_window_key_field then window field
     input_df = (
-        (
-            input_df.groupby(key_fields)
-            .apply(
-                get_timestamps_for_session_window,
-                key_fields=key_fields,
-                ts_field=ts_field,
-                gap=gap,
-            )
-            .reset_index()
-        )
-        .loc[:, cols]
-        .assign(
-            **{
-                ts_field: lambda x: x[ts_field].astype(
-                    pd.ArrowDtype(pa.timestamp("ns", "UTC"))
-                )
-            }
+        input_df.groupby(key_fields_without_window)
+        .apply(
+            get_timestamps_for_session_window,
+            key_fields_without_window=key_fields_without_window,
+            ts_field=ts_field,
+            gap=gap,
+            is_window_key_field=is_window_key_field,
         )
+        .reset_index()
+    )
+    input_df = input_df.loc[:, cols]
+    input_df[ts_field] = cast_col_to_arrow_dtype(
+        input_df[ts_field], get_timestamp_data_type()
     )
 
-    fennel_internal_field = "__fennel_internal_field__"
+    key_fields = key_fields_without_window.copy()
 
-    # Getting all the unique sessions against all the keys
+    # Cast key window field to first arrow dtype and then to frozen dict dtype
+    if is_window_key_field:
+        input_df[WINDOW_FIELD_NAME] = cast_col_to_arrow_dtype(
+            input_df[WINDOW_FIELD_NAME], get_window_data_type()
+        ).apply(lambda x: frozendict(x))
+        key_fields.append(WINDOW_FIELD_NAME)
+
+    fennel_internal_ts_field = "__fennel_internal_ts_field__"
+    fennel_internal_window_field = "__fennel_internal_window_field__"
+
```

Comment:
should these go on the top of the file ?

Reply:
moved.